File: app/src/main/java/me/zhanghai/android/files/filelist/CreateArchiveDialogFragment.java
/*
 * Copyright (c) 2019 Hai Zhang <dreaming.in.code.zh@gmail.com>
 * All Rights Reserved.
 */

package me.zhanghai.android.files.filelist;

import android.app.Dialog;
import android.os.Bundle;
import android.widget.RadioGroup;

import org.apache.commons.compress.archivers.ArchiveStreamFactory;
import org.apache.commons.compress.compressors.CompressorStreamFactory;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.Set;

import androidx.annotation.LayoutRes;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import butterknife.BindView;
import me.zhanghai.android.files.R;
import me.zhanghai.android.files.util.CollectionUtils;
import me.zhanghai.android.files.util.FragmentUtils;

public class CreateArchiveDialogFragment extends FileNameDialogFragment {

    private static final String KEY_PREFIX = CreateArchiveDialogFragment.class.getName() + '.';

    private static final String EXTRA_FILES = KEY_PREFIX + "FILES";

    @BindView(R.id.type_group)
    RadioGroup mTypeGroup;

    private Set<FileItem> mExtraFiles;

    @NonNull
    public static CreateArchiveDialogFragment newInstance(@NonNull Set<FileItem> files) {
        //noinspection deprecation
        CreateArchiveDialogFragment fragment = new CreateArchiveDialogFragment();
        FragmentUtils.getArgumentsBuilder(fragment)
                .putParcelableArrayList(EXTRA_FILES, new ArrayList<>(files));
        return fragment;
    }

    public static void show(@NonNull Set<FileItem> files, @NonNull Fragment fragment) {
        CreateArchiveDialogFragment.newInstance(files)
                .show(fragment.getChildFragmentManager(), null);
    }

    /**
     * @deprecated Use {@link #newInstance(Set)} instead.
     */
    public CreateArchiveDialogFragment() {}

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        mExtraFiles = new HashSet<>(getArguments().getParcelableArrayList(EXTRA_FILES));
    }

    @NonNull
    @Override
    public Dialog onCreateDialog(@Nullable Bundle savedInstanceState) {
        Dialog dialog = super.onCreateDialog(savedInstanceState);
        if (savedInstanceState == null) {
            if (mExtraFiles.size() == 1) {
                FileItem file = CollectionUtils.first(mExtraFiles);
                String name = file.getPath().getFileName().toString();
                mNameEdit.setText(name);
                mNameEdit.setSelection(0, name.length());
            }
        }
        return dialog;
    }

    @LayoutRes
    @Override
    protected int getTitleRes() {
        return R.string.file_create_archive_title;
    }

    @Override
    protected int getLayoutRes() {
        return R.layout.create_archive_dialog;
    }

    @NonNull
    @Override
    protected String getName() {
        int typeId = mTypeGroup.getCheckedRadioButtonId();
        String extension;
        switch (typeId) {
            case R.id.type_zip:
                extension = "zip";
                break;
            case R.id.type_tar_xz:
                extension = "tar.xz";
                break;
            case R.id.type_seven_z:
                extension = "7z";
                break;
            default:
                throw new AssertionError(typeId);
        }
        return super.getName() + '.' + extension;
    }

    @Override
    protected void onOk(@NonNull String name) {
        int typeId = mTypeGroup.getCheckedRadioButtonId();
        String archiveType;
        String compressorType;
        switch (typeId) {
            case R.id.type_zip:
                archiveType = ArchiveStreamFactory.ZIP;
                compressorType = null;
                break;
            case R.id.type_tar_xz:
                archiveType = ArchiveStreamFactory.TAR;
                compressorType = CompressorStreamFactory.XZ;
                break;
            case R.id.type_seven_z:
                archiveType = ArchiveStreamFactory.SEVEN_Z;
                compressorType = null;
                break;
            default:
                throw new AssertionError(typeId);
        }
        getListener().archive(mExtraFiles, name, archiveType, compressorType);
    }

    @NonNull
    @Override
    protected Listener getListener() {
        return (Listener) requireParentFragment();
    }

    public interface Listener extends FileNameDialogFragment.Listener {
        void archive(@NonNull Set<FileItem> files, @NonNull String name,
                     @NonNull String archiveType, @Nullable String compressorType);
    }
}

File: app/src/main/java/me/zhanghai/android/files/filelist/CreateArchiveDialogFragment.kt
/*
 * Copyright (c) 2019 Hai Zhang <dreaming.in.code.zh@gmail.com>
 * All Rights Reserved.
 */

package me.zhanghai.android.files.filelist

import android.app.Dialog
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.widget.EditText
import android.widget.RadioGroup
import androidx.annotation.StringRes
import androidx.core.view.isGone
import androidx.fragment.app.Fragment
import com.google.android.material.textfield.TextInputEditText
import com.google.android.material.textfield.TextInputLayout
import kotlinx.parcelize.Parcelize
import me.zhanghai.android.files.R
import me.zhanghai.android.files.databinding.CreateArchiveDialogBinding
import me.zhanghai.android.files.databinding.NameDialogNameIncludeBinding
import me.zhanghai.android.files.util.ParcelableArgs
import me.zhanghai.android.files.util.args
import me.zhanghai.android.files.util.putArgs
import me.zhanghai.android.files.util.setTextWithSelection
import me.zhanghai.android.files.util.show
import me.zhanghai.android.files.util.takeIfNotEmpty
import me.zhanghai.android.libarchive.Archive

class CreateArchiveDialogFragment : FileNameDialogFragment() {
    private val args by args<Args>()

    override val binding: Binding
        get() = super.binding as Binding

    override val listener: Listener
        get() = super.listener as Listener

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        val dialog = super.onCreateDialog(savedInstanceState)

        if (savedInstanceState == null) {
            val files = args.files
            var name: String? = null
            if (files.size == 1) {
                name = files.single().path.fileName.toString()
            } else {
                val parent = files.mapTo(mutableSetOf()) { it.path.parent }.singleOrNull()
                if (parent != null && parent.nameCount > 0) {
                    name = parent.fileName.toString()
                }
            }
            name?.let { binding.nameEdit.setTextWithSelection(it) }
        }
        binding.typeGroup.setOnCheckedChangeListener { _, _ -> updatePasswordLayoutVisibility() }
        updatePasswordLayoutVisibility()
        return dialog
    }

    @StringRes
    override val titleRes: Int = R.string.file_create_archive_title

    override fun onInflateBinding(inflater: LayoutInflater): NameDialogFragment.Binding =
        Binding.inflate(inflater)

    override val name: String
        get() {
            val extension = when (val checkedId = binding.typeGroup.checkedRadioButtonId) {
                R.id.zipRadio -> "zip"
                R.id.tarXzRadio -> "tar.xz"
                R.id.sevenZRadio -> "7z"
                else -> throw AssertionError(checkedId)
            }
            return "${super.name}.$extension"
        }

    private val isPasswordSupported: Boolean
        get() = when (val checkedId = binding.typeGroup.checkedRadioButtonId) {
            R.id.zipRadio -> true
            R.id.tarXzRadio, R.id.sevenZRadio -> false
            else -> throw AssertionError(checkedId)
        }

    private fun updatePasswordLayoutVisibility() {
        binding.passwordLayout.isGone = !isPasswordSupported
    }

    override fun onOk(name: String) {
        val (format, filter) = when (val checkedId = binding.typeGroup.checkedRadioButtonId) {
            R.id.zipRadio -> Archive.FORMAT_ZIP to Archive.FILTER_NONE
            R.id.tarXzRadio -> Archive.FORMAT_TAR to Archive.FILTER_XZ
            R.id.sevenZRadio -> Archive.FORMAT_7ZIP to Archive.FILTER_NONE
            else -> throw AssertionError(checkedId)
        }
        val password = if (isPasswordSupported) {
            binding.passwordEdit.text!!.toString().takeIfNotEmpty()
        } else {
            null
        }
        listener.archive(args.files, name, format, filter, password)
    }

    companion object {
        fun show(files: FileItemSet, fragment: Fragment) {
            CreateArchiveDialogFragment().putArgs(Args(files)).show(fragment)
        }
    }

    @Parcelize
    class Args(val files: FileItemSet) : ParcelableArgs

    protected class Binding private constructor(
        root: View,
        nameLayout: TextInputLayout,
        nameEdit: EditText,
        val typeGroup: RadioGroup,
        val passwordLayout: TextInputLayout,
        val passwordEdit: TextInputEditText
    ) : NameDialogFragment.Binding(root, nameLayout, nameEdit) {
        companion object {
            fun inflate(inflater: LayoutInflater): Binding {
                val binding = CreateArchiveDialogBinding.inflate(inflater)
                val bindingRoot = binding.root
                val nameBinding = NameDialogNameIncludeBinding.bind(bindingRoot)
                return Binding(
                    bindingRoot, nameBinding.nameLayout, nameBinding.nameEdit, binding.typeGroup,
                    binding.passwordLayout, binding.passwordEdit
                )
            }
        }
    }

    interface Listener : FileNameDialogFragment.Listener {
        fun archive(files: FileItemSet, name: String, format: Int, filter: Int, password: String?)
    }
}
